File: frontend/src/types/emotion.ts
export type EmotionLabel =
  | "anger"
  | "disgust"
  | "fear"
  | "joy"
  | "neutral"
  | "sadness"
  | "surprise";

export type Channel = "text" | "voice" | "face";

export type EmotionScores = Record<EmotionLabel, number>;

export interface EmotionPrediction {
  label: EmotionLabel;
  confidence: number;
  scores: EmotionScores;
}

export interface LiveAudioAnalysis {
  transcript: string;
  audio_prediction: EmotionPrediction | null;
  text_prediction: EmotionPrediction | null;
}

/** Mirrors backend/app/schemas/emotion.py. */
export interface FusedPrediction {
  label: EmotionLabel;
  /** Attenuated by divergence. This is the value to display. */
  confidence: number;
  raw_confidence: number;
  attenuation: number;
  scores: EmotionScores;
  weights: Partial<Record<Channel, number>>;
}

export interface PairDivergence {
  channels: Channel[];
  jensen_shannon: number;
  cosine_distance: number;
}

export interface ConflictAnalysis {
  status: "conflict" | "aligned" | "insufficient_channels";
  channels_compared: Channel[];
  pairs: PairDivergence[];
  max_divergence: number | null;
  mean_divergence: number | null;
  most_divergent_pair: Channel[] | null;
  threshold: number;
  conflict_detected: boolean;
}

export interface FusionAnalysis {
  fused: FusedPrediction | null;
  channels: Partial<Record<Channel, EmotionScores>>;
  conflict: ConflictAnalysis;
}

/** Mirrors backend/app/schemas/session.py. */
export interface ReadingIn {
  t: string;
  channel: Channel;
  label: EmotionLabel;
  confidence: number;
  scores: EmotionScores;
}

export interface FusedReadingIn {
  t: string;
  label: EmotionLabel;
  confidence: number;
  raw_confidence: number;
  attenuation: number;
  max_divergence: number | null;
  conflict: boolean;
  scores: EmotionScores;
}

export interface ReadingBatch {
  readings: ReadingIn[];
  fused: FusedReadingIn[];
}

export interface SessionOut {
  id: string;
  started_at: string;
  ended_at: string | null;
}

export interface SessionSummary {
  session_id: string;
  n_readings: number;
  n_fused_readings: number;
  mean_valence: number | null;
  conflict_rate: number | null;
  dominant_label: EmotionLabel | null;
  channel_counts: Partial<Record<Channel, number>>;
  computed_at: string;
}

File: backend/app/schemas/emotion.py
from pydantic import BaseModel, Field, field_validator

CANONICAL_EMOTIONS = (
    "anger",
    "disgust",
    "fear",
    "joy",
    "neutral",
    "sadness",
    "surprise",
)

CHANNELS = ("text", "voice", "face")


class TextPredictionRequest(BaseModel):
    text: str = Field(max_length=5000)


class EmotionPrediction(BaseModel):
    label: str
    confidence: float
    scores: dict[str, float]


class LiveAudioAnalysis(BaseModel):
    transcript: str
    audio_prediction: EmotionPrediction | None
    text_prediction: EmotionPrediction | None


class HealthResponse(BaseModel):
    status: str
    models: dict[str, bool]


class ConflictRequest(BaseModel):
    """Latest score vector per channel. Any two channels are enough to compare."""

    text: dict[str, float] | None = None
    voice: dict[str, float] | None = None
    face: dict[str, float] | None = None

    @field_validator("text", "voice", "face")
    @classmethod
    def validate_channel_scores(cls, scores: dict[str, float] | None) -> dict[str, float] | None:
        if scores is None:
            return None
        unknown = set(scores) - set(CANONICAL_EMOTIONS)
        if unknown:
            raise ValueError(f"Unknown emotion labels: {', '.join(sorted(unknown))}.")
        if any(value < 0 for value in scores.values()):
            raise ValueError("Emotion scores cannot be negative.")
        if sum(scores.values()) <= 0:
            raise ValueError("Emotion scores must contain at least one positive value.")
        return scores

    def channels(self) -> dict[str, dict[str, float]]:
        present = {"text": self.text, "voice": self.voice, "face": self.face}
        return {name: scores for name, scores in present.items() if scores is not None}


class FusedPrediction(BaseModel):
    """The composite reading. `confidence` is the value to display."""

    label: str
    confidence: float
    raw_confidence: float
    attenuation: float
    scores: dict[str, float]
    weights: dict[str, float]


class PairDivergence(BaseModel):
    channels: list[str]
    jensen_shannon: float
    cosine_distance: float


class ConflictAnalysis(BaseModel):
    """Divergence between channels. Never a claim about a person's inner state."""

    status: str
    channels_compared: list[str]
    pairs: list[PairDivergence]
    max_divergence: float | None
    mean_divergence: float | None
    most_divergent_pair: list[str] | None
    threshold: float
    conflict_detected: bool


class FusionAnalysis(BaseModel):
    """Fused headline plus the per-channel components it was built from."""

    fused: FusedPrediction | None
    channels: dict[str, dict[str, float]]
    conflict: ConflictAnalysis
